---
title: "Tutorial 2: Quality control and Preprocessing of the snRNA-seq Data"
author: "Thodoris Koutsandreas, Umran Yaman, Iguaracy Pinheiro-de-Sousa"
#date: 'Compiled: `r Sys.Date()`'  
output: 
  html_document:
    toc: TRUE
    code_folding: show
    theme: united
    df_print: kable
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  warning = FALSE,
  error = TRUE,
  echo = TRUE,
  message = FALSE,
  fig.width = 8,
  time_it = TRUE,
  include = TRUE
)
```



```{r}
suppressPackageStartupMessages({
  library(ggplot2)
  library(Seurat)
  library(DoubletFinder)
})
```

<br>
<br>

### 1. Loading of the snRNA-seq Dataset
The initial dataset contains snRNA-seq data for more than 414k cells and it could 
be found [here](https://compbio.mit.edu/ad_epigenome/). Similarly to the 
scATAC-seq data, a subset of this dataset has been created, down sampling each 
cell type separately in order to have a maximum of 3500 cells. The subset() 
function was used for this task.

```{r}
snRNA_seurat_obj <- readRDS("../data/initial_data/downsampled_snRNA_seurat_obj.rds")
DefaultAssay(snRNA_seurat_obj) <- "RNA"
```

```{r}
# Number of cells per cell type.
table(snRNA_seurat_obj@meta.data$major.celltype)
```

```{r}
# Number of cells per pathological state.
table(snRNA_seurat_obj@meta.data$Pathology)
```

<br>
<br>

### 2. Quality Control for Cells
The QC task involves several steps to examine the quality of the data and 
remove low quality cells and features (genes). Different figures 
could help for this task, visualizing the distribution of various meta data 
features of cells (library sizes, mitochondrial and ribosomal genes percentage, 
etc).

<br>

#### Visualization of the distribution of cells metadata
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
features = c("n_genes","n_counts","pct_mito","pct_ribo")
# VlnPlot() is a function in Seurat
VlnPlot(snRNA_seurat_obj, 
        features = features,
        ncol = 4, pt.size = 0.0)
```

<br>

#### Visualization of the log2-transformed library sizes and expressed genes number in order to identify low-quality cells
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
# Calculation of the minimum threshold for the library sizes (log2-transformed)
values <- log2(snRNA_seurat_obj@meta.data$n_counts)
median_value <- median(values)
threshold_for_library_size <- median_value - 3* median(abs(median_value-values))
bool1 <- values <= threshold_for_library_size 

# Calculation of the minimum threshold for the number of expressed genes (log2-transformed)
values <- log2(snRNA_seurat_obj@meta.data$n_genes)
median_value <- median(values)
threshold_for_n_genes <- median_value - 3* median(abs(median_value-values))
bool2 <- values <= threshold_for_n_genes

# Scatter plot for the above log-transformed values
ggplot(data = snRNA_seurat_obj@meta.data, aes(x = log2(n_counts), y = log2(n_genes))) +
  geom_point(aes(colour = (bool1 | bool2)), 
             shape = 19, alpha = 0.5, size=0.1) +
  scale_colour_discrete(type = c("#0059b3", "#b30059"), name = 'Low quality cell: ') +
  labs(x = 'log2(library size)', y = 'log2(number of expressed genes)') +
  theme(axis.title.x = element_text(size = 14),
        axis.title.y = element_text(size = 14),
        axis.text.x = element_text(size = 12),
        axis.text.y = element_text(size = 12),
        panel.background = element_rect(fill = '#ffffff'),
        panel.grid = element_line(colour = '#D5D5D7'),
        panel.border = element_rect(colour = "#000000", fill = NA, linewidth = 0.5),
        legend.key.size = unit(0.1, 'cm'),
        legend.key = element_rect(fill = '#ffffff'),
        legend.justification = "center",
        legend.position = "bottom")+
  guides(color = guide_legend(override.aes = list(linewidth = 0.5, size=8), 
                              nrow = 1, title.position="left", title.vjust = 0.5)) 
```

<br>

#### Visualization of percentages of mitochondrial and ribosomal genes
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
tmp_df <- rbind(
  data.frame('percentage'=snRNA_seurat_obj@meta.data$pct_mito,
             'criterion'='mitochondrial genes'),
  data.frame('percentage'=snRNA_seurat_obj@meta.data$pct_ribo,
             'criterion'='ribosomal genes')
)

ggplot(tmp_df, aes(x=percentage)) + 
  geom_histogram(aes(y=..density..), binwidth=0.01, colour="black", fill="white") +
    facet_grid(criterion ~ ., scales = "free") +
  labs(title="Percentages of mitochondrial and ribosomal genes") +
  theme(axis.title = element_text(size=20),
        axis.title.y = element_text(size=16),
        axis.title.x = element_text(size=16),
        axis.text = element_text(size=12))
```

<br>

#### Removal of low-quality cells
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
# Before the QC filtering
print(dim(snRNA_seurat_obj))

# Subset
snRNA_qc_seurat_obj <- subset(snRNA_seurat_obj,
                              n_counts > 2**threshold_for_library_size &
                              n_genes > 2**threshold_for_n_genes &
                              pct_mito < 10 &
                              pct_ribo < 10)

# After the QC filtering
print(dim(snRNA_qc_seurat_obj))
```

<br>
<br>

### 3. Quality Control for Genes
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
# Average expression of genes (this is the rowMeans() of counts matrix)
average_expressions <- rowMeans(snRNA_qc_seurat_obj@assays$RNA$counts)
tmp_df <- data.frame(average_expression = average_expressions)

ggplot(tmp_df, aes(x=log10(average_expression))) + 
  geom_histogram(aes(y=..density..), binwidth=0.25, colour="black", fill="white") +
  xlim(-5,5)+
  labs(title="Log10(gene average expression)") +
  theme(axis.title = element_text(size=20),
        axis.title.y = element_text(size=16),
        axis.title.x = element_text(size=16),
        axis.text = element_text(size=12))
```

<br>

#### Removal of low-expressed genes
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
# Before the QC filtering
print(dim(snRNA_qc_seurat_obj))

# Subset
selected_features <- names(average_expressions[average_expressions > 0.01])
snRNA_qc_seurat_obj <- subset(snRNA_qc_seurat_obj, features = selected_features)

# After the QC filtering
print(dim(snRNA_qc_seurat_obj))


# Remove old object and clean-up the space
rm(snRNA_seurat_obj)
gc()
```

<br>
<br>

### 4. Data Normalization
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
# The normalized counts matrix will be stores in the 'data' slot.
snRNA_qc_seurat_obj <- NormalizeData(snRNA_qc_seurat_obj, assay='RNA', 
                                     normalization.method = "LogNormalize", 
                                     scale.factor = 10000)
dim(snRNA_qc_seurat_obj@assays$RNA@data)

# Different normalization methods:

# 1. LogNormalize: Feature counts for each cell are divided by the total counts 
# for that cell and multiplied by the scale.factor. This is then natural-log 
# transformed using log1p.

# 2. CLR: Applies a centered log ratio transformation.

# 3. RC: Relative counts. Feature counts for each cell are divided by the total 
# counts for that cell and multiplied by the scale.factor. No log-transformation 
# is applied. For counts per million (CPM) set scale.factor = 1e6.

```

<br>
<br>

### 5. Calculation of Cell Cycle Scores
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
# cc.genes and CellCycleScoring() are used from Seurat
cc.genes.updated.2019$s.genes <- UpdateSymbolList(symbols = cc.genes.updated.2019$s.genes)
cc.genes.updated.2019$g2m.genes <- UpdateSymbolList(symbols = cc.genes.updated.2019$g2m.genes)

snRNA_qc_seurat_obj <- CellCycleScoring(snRNA_qc_seurat_obj, 
                                        s.features = cc.genes.updated.2019$s.genes, 
                                        g2m.features = cc.genes.updated.2019$g2m.genes, 
                                        set.ident = TRUE)
```

<br>

#### Visualization of cell cycle scores disitrbutions per phase
```{r}
VlnPlot(snRNA_qc_seurat_obj, features = c("S.Score", "G2M.Score"), 
        group.by = 'Phase', ncol = 2, pt.size = 0)
```


```{r}
ggplot(snRNA_qc_seurat_obj@meta.data, aes(x=S.Score, y=G2M.Score, col=Phase))+
  geom_point(size=0.1)
```

<br>
<br>

### 6. Identification of the Most Variable Genes
```{r, tidy='styler', tidy.opts=list(strict=FALSE)}
snRNA_qc_seurat_obj <- FindVariableFeatures(snRNA_qc_seurat_obj,
                                            selection.method = "vst", 
                                            nfeatures = 2000, verbose = F)

# vst: First, fits a line to the relationship of log(variance) and log(mean)
# using local polynomial regression (loess). Then standardizes the feature 
# values using the observed mean and expected variance (given by the fitted line). 
# Feature variance is then calculated on the standardized values after clipping 
# to a maximum (see clip.max parameter).

top10 <- head(VariableFeatures(snRNA_qc_seurat_obj), 10)
p <- VariableFeaturePlot(snRNA_qc_seurat_obj,  pt.size = 0.1, cols = c("black", "red"))
LabelPoints(plot = p, points = top10, repel = TRUE, xnudge = 0, ynudge = 0)
```

<br>
<br>

### 7. Further Data Transformation - Scaling & Centering
Regress out the read depth, mitochondrial percentage and cell cycling genes.
```{r}
variables <- c("n_genes", "pct_mito","S.Score", "G2M.Score")
snRNA_qc_seurat_obj = ScaleData(snRNA_qc_seurat_obj, vars.to.regress = variables, 
                                verbose = FALSE)
```

<br>
<br>

### 8. Dimensionality Reduction: PCA & UMAP

<br>

#### Principal Components Analysis (PCA)
PCA works by transforming the original gene expression data into a new set of 
orthogonal variables, called principal components (PCs). These PCs are linear 
combinations of the original genes and are ordered such that the first PC 
captures the maximum amount of variance in the data, followed by the second 
PC capturing the maximum remaining variance orthogonal to the first PC, and so on.
```{r}
# PCA runs on the scaled data
snRNA_qc_seurat_obj <- RunPCA(snRNA_qc_seurat_obj, npcs = 50, 
                              features = VariableFeatures(object = snRNA_qc_seurat_obj))
```

<br>

#### Explained variance (%) for each Principal Component (first 30)
```{r}
pca <- snRNA_qc_seurat_obj@reductions$pca
pca_matrix <- pca@cell.embeddings
total_variance <- pca@misc$total.variance
explained_variance <- 100 * (pca@stdev)^2 / total_variance

tmp_df <- data.frame(PC=seq(1,100), 
                     explained_variance=explained_variance)

ggplot(tmp_df[1:30,], aes(x=PC, y=explained_variance)) + 
  geom_bar(stat = "identity") +
  labs(x='Principal Component', y ='Explained Variance (%)') +
  theme(axis.title = element_text(size=16),
        axis.text = element_text(size=14))
```

<br>

#### Visualization of cells using the first two Principal Components
```{r}
DimPlot(snRNA_qc_seurat_obj, reduction='pca', group.by='major.celltype')
```

<br>

#### Uniform Manifold Approximation and Projection (UMAP)
UMAP is another dimensionality  reduction technique, which is traditionally 
used in single cell analysis in order to embed the expression profile of cells 
into a two-dimensional space. The distances between the cells in this new 
space approximate the original ones.  
```{r}
snRNA_qc_seurat_obj <- RunUMAP(snRNA_qc_seurat_obj, dims = 1:5, verbose = FALSE)
```

<br>

#### Visualization of cells in the UMAP space
```{r}
DimPlot(snRNA_qc_seurat_obj, reduction='umap', group.by='major.celltype')
```

<br>
<br>

### 9. Removal of Potential Doublets

Doublets are artificial cells, generated from the integration of two cells. 
They could be generated due to errors in cell sorting or capturing. 
As they are artificial libraries we want to detect and filter them out.
```{r}
expected_ratio <- round(ncol(snRNA_qc_seurat_obj) * 0.04)  # expect 4% doublets

# The doubletFinder() function from DoubletFinder is used
snRNA_qc_seurat_obj <- doubletFinder(snRNA_qc_seurat_obj, pN = 0.1, pK = 0.09, 
                                     nExp = expected_ratio, PCs = 1:10)

# name of the doubletFinder prediction can change, so extract the correct column 
# name from the meta.data.
colnames <- colnames(snRNA_qc_seurat_obj@meta.data)
doublets_indication_col = colnames[grepl("DF.classification", colnames)]
```

<br>

#### Visualization of cells according to doublets identity
```{r}
DimPlot(snRNA_qc_seurat_obj, group.by = doublets_indication_col, pt.size = 0.1) + NoAxes()
```

<br>

#### Subset of the Seurat object
```{r}
# Keep only singlets
snRNA_qc_seurat_obj = snRNA_qc_seurat_obj[, snRNA_qc_seurat_obj@meta.data[, doublets_indication_col] == "Singlet"]
dim(snRNA_qc_seurat_obj)
```

<br>
<br>

### 10. Saving of the QC and Preprocessed Dataset
```{r}
saveRDS(snRNA_qc_seurat_obj, file = "../data/snRNA_qc_seurat_obj.rds")
```


```{r}
sessionInfo()
```


