---
title: "Tutorial 6. TF Activity Analysis Using the snATAC-seq Data"
author: "Iguaracy Souza, Thodoris Koutsandreas, and Umran Yaman"
#date: 'Compiled: `r Sys.Date()`'  
output: 
  html_document:
    toc: TRUE
    code_folding:  "show"
    theme: united
    df_print: kable
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  warning = FALSE,
  error = TRUE,
  echo = TRUE,
  message = FALSE,
  fig.width = 8,
  time_it = TRUE
)
```


```{r, include=FALSE}
suppressPackageStartupMessages({
  library(Signac)
  library(Seurat)
  library(BSgenome.Hsapiens.UCSC.hg38)
  library(JASPAR2020)
  library(TFBSTools)
  library(motifmatchr)
  library(ggpubr)
  library(ggseqlogo)
})
```

### 1. Loading of the snATAC-seq data
```{r}
snATAC_qc_seurat_obj <- readRDS("../data/downsampled_snATAC_qc_seurat_obj.rds")
DefaultAssay(snATAC_qc_seurat_obj) <- 'ATAC'
Idents(snATAC_qc_seurat_obj) <- snATAC_qc_seurat_obj$Celltype1
```

<br>
<br>

### 2. Binding Sites of Transcription Factors

<br>

The binding sites for transcription factors (in the form of motifs) will be 
retrieved from the JASPAR database. JASPAR contains curated, non-redundant and 
experimentally defined sets of transcription factor binding sites for eukaryotes.
This information will be used to detect the sequence ranges on the genome which 
match with these motifs, identifying in this way the binding sites of TFs and 
the genes that correspond to the matched sequence ranges.
```{r}
# Get a position frequency matrices (PFMs) object from the JASPAR database
# 9606 is the NCBI Taxonomy ID for Homo sapiens
position_freq_matrices  <- getMatrixSet(x = JASPAR2020, 
                                        opts = list(species = 9606, 
                                                    all_versions = FALSE)
                                        )

# The following command scans the DNA sequence of each peak (sequence range), 
# using the genome data, to identify for the presence of each motif and store 
# that information in the the Seurat object.
snATAC_qc_seurat_obj <- AddMotifs(snATAC_qc_seurat_obj, 
                                  pfm = position_freq_matrices,
                                  genome = BSgenome.Hsapiens.UCSC.hg38)
```

<br>

Plot of TF binding sites (motifs).
```{r, include=TRUE, tidy='styler', tidy.opts=list(strict=FALSE), fig.height = 3, fig.width = 8}
motifs <- Motifs(snATAC_qc_seurat_obj)

motif_id <- ConvertMotifID(motifs, name = 'SMAD3')
p1 <- MotifPlot(snATAC_qc_seurat_obj, motifs = motif_id)

motif_id <- ConvertMotifID(motifs, name = 'NFKB1')
p2 <- MotifPlot(snATAC_qc_seurat_obj, motifs = motif_id)

ggarrange(p1, p2, ncol=1, nrow=2)
```

<br>

Find the matched sequence ranges on the genome for a given TF.
```{r, include=TRUE, tidy='styler', tidy.opts=list(strict=FALSE), fig.height = 3, fig.width = 8}
motif_id <- ConvertMotifID(motifs, name = 'SMAD3')
ranges <- motifs@data[,motif_id]
matched_ranges <- ranges[ranges == TRUE]

print(paste('Sequences ranges in genome:', length(ranges), sep=' '))
print(paste('Matched sequences for SMAD3:', length(matched_ranges), sep=' '))
```

<br>
<br>

### 3. Differential Transcription Factors Activity Analysis

<br>

The analysis will be performed for each pathological condition Thus, the dataset 
should be separated according to this factor.
```{r}
snATAC_by_condition <- SplitObject(snATAC_qc_seurat_obj, split.by = "Pathology")
```

<br>

Firstly the FindMarkers() function will be used to perform differential analysis
in order to identify peaks that are significantly differentially accessible 
between cell types for each condition.
```{r}
peaks_list <- list()

for (condition in names(snATAC_by_condition)) {
  
  condition_seurat_obj <- snATAC_by_condition[[condition]] 
  cell_types <- unique(condition_seurat_obj@meta.data$Celltype1)

  # Loop through each cell type
  for (cell_type in cell_types) {
    # Check the number of cells in this cell type
    num_cells <- sum(condition_seurat_obj@meta.data$Celltype1 == cell_type)
    if (num_cells < 10) {
      next  # Skip to the next cell type
    }
    # Perform differential expression analysis
    peaks_df <- FindMarkers(condition_seurat_obj, ident.1 = cell_type, 
                            only.pos = TRUE, test.use = 'LR',
                            min.pct = 0.3, latent.vars = 'nCount_ATAC')
    peaks_df <- peaks_df[(peaks_df$p_val_adj < 0.05 & peaks_df$avg_log2FC > 0.5),]
    if (dim(peaks_df)[1] > 1500) {
        # Be stricter
        peaks_df <- peaks_df[(peaks_df$p_val_adj < 0.05 & peaks_df$avg_log2FC > 1),]
    }
    peaks_list[[paste(condition, cell_type, sep = "_")]] <- peaks_df
    print(paste(condition, cell_type, dim(peaks_df)[1]))
  }
}
```

<br>

Secondly, the FindMotifs() function will be used to identify those motifs that 
are over-represented in the set of differentially accessible peaks in each condition
and cell type. In other words, this function will reveal those TFs whose activity
is probably deregulated, based on the differentiated accessibility on the respective
sites on the genome.
```{r}
enriched_motifs <- list()

for (key in names(peaks_list)) {
  
  condition <- paste(strsplit(key, split='_')[[1]][1:2], collapse='_')
  condition_seurat_obj <- snATAC_by_condition[[condition]] 
  background <- rownames(condition_seurat_obj)
  da_peaks <- rownames(peaks_list[[key]])
  
  motifs_df <- FindMotifs(
    object = condition_seurat_obj, 
    features = da_peaks,
    background = background
    )
  motifs_df <- motifs_df[(motifs_df$p.adjust < 0.1 & motifs_df$fold.enrichment > 2),]
  enriched_motifs[[key]] <- motifs_df
}

# Save the enriched_motifs list in the 'results' folder
saveRDS(enriched_motifs, '../results/tf_results.rds')
```

